cmd/horologium/main.go

```diff
@@ -199,14 +202,27 @@ func sync(prowJobClient ctrlruntimeclient.Client, cfg *config.Config, cr cronCli
 			}
 			continue
 		}
-		if !shouldTrigger {
+		if !shouldTrigger && p.RetriggerFailedRun == nil {
 			logger.WithFields(logrus.Fields{
 				"previous-found": previousFound,
 				"name":           p.Name,
 				"job":            p.JobBase.Name,
 			}).Debug("Trigger time has not yet been reached.")
 		}
+		run := false
+
+		// Proceed only if the run deviates from the normal flow.
+		if !shouldTrigger {
+			run = shouldTriggerFailedRun(j, p, now, logger, previousFound)
+		}
 		if !previousFound || shouldTrigger {
+			run = true
+		}
+		if run {
+			if p.RetriggerFailedRun != nil && p.Labels == nil {
+				p.Labels = make(map[string]string)
+				p.Labels[kube.ReRunLabel] = strconv.Itoa(1)
```

Comment:
L224 looks like something that needs to be done even when `p.Labels` was not nil

Reply:
Actually, one thing I noticed - `p.Labels` is the config structure. Let's not modify the config structure, but create  a new map from it (=populate defaults from the config) and add the controller's labels in there.